toroid.py

```diff
@@ -0,0 +1,304 @@
+'''
+
+
+class Toroid:
+    '''A class representing a z-axis oriented, origin-centered elliptical toroid in terms of: 
+        r, the major radius (along xy plane); 
+        a, the radius of the rotated ellipse which is parallel to the major radius (xy plane); and 
+        b, the ellipse radius orthogonal to the major radius (z axis). 
+        
+    Class objects also contain three methods for the three surface functions: distance from point
+    along ray, surface sense at point, and surface normal at point.
+
+    Note
+    ----
+    The class also contains a secondary representation of its properties in terms of 3 parameters
+    p, A, and B, for the purpose of solving ray intersection. However, accessing these parameters
+    directly is highly discouraged, as their precise implementation is non-final.
+    
+    '''
+    def __init__(self, r: float, a: float, b: float):
+        '''
+        Parameters
+        ----------
+        r : float
+            The radius from the origin of the toroid to the center of the revolved ellipse, or the "Major Radius", along the xy plane
+        a : float
+            The radius of the revolved ellipse aligned with the major radius, or the "parallel minor radius", along the xy plane
+        b : float
+            The radius of the revolved ellipse orthogonal to the major radius, or the "orthogonal minor radius", aligned with the z-axis
+
+        Attributes
+        ----------
+        TODO: Do you define the effective attributes that the user sees here, or include the private facing ones? I assume the former, in which case it should basically be identical
+        '''
+        self._r = r
+        self._a = a
+        self._b = b
+
+        # From Graphics Gems, form which is more convenient for solving ray intersection
+        self._p = (a*a) / (b*b)
+        self._A = 4*r*r
+        self._B = r*r  - a*a
+
+    @property
+    def r(self):
+        return self._r
+
+    @property
+    def a(self):
+        return self._a
+
+    @property
+    def b(self):
+        return self._b
+
+    @property
+    def p(self):
+        return self._p
+
+    @property
+    def A(self):
+        return self._A
+
+    @property
+    def B(self):
+        return self._B
+    
+    @r.setter
+    def r(self, value):
+        self._r = value
+        r = self.r
+        a = self.a
+        b = self.b
+        self._p = (a*a) / (b*b)
+        self._A = 4*r*r
+        self._B = r*r  - a*a
+
+    @a.setter
+    def a(self, value):
+        self._a = value
+        r = self.r
+        a = self.a
+        b = self.b
+        self._p = (a*a) / (b*b)
+        self._A = 4*r*r
+        self._B = r*r  - a*a
+    
+    @b.setter
+    def b(self, value):
+        self._b = value
+        r = self.r
+        a = self.a
+        b = self.b
+        self._p = (a*a) / (b*b)
+        self._A = 4*r*r
+        self._B = r*r  - a*a
+
+    def _ray_intersection_polynomial(
+        self, ray_src: Iterable[float], ray_dir: Iterable[float]
```

Comment:
rename ray_src and ray_dir to pos and dir, here and throughout (conventional in ORANGE)